rflx/pyrflx/message.py

```diff
@@ -0,0 +1,168 @@
+from typing import Any, List, Mapping
+
+import rflx.model as model
+from rflx.expression import TRUE, UNDEFINED, Add, Expr, First, Length, Name, Number, Variable
+
+from .typevalue import OpaqueValue, ScalarValue, TypeValue
+
+
+class Field:
+    name: str
+    typeval: TypeValue
+    first: Expr
+    length: Expr
+
+    def __init__(self, name: str) -> None:
+        self.name = name
+        self.typeval = TypeValue.construct(model.Opaque())
+        self.first = UNDEFINED
+        self.length = UNDEFINED
+
+    def __eq__(self, other: object) -> bool:
+        if isinstance(other, self.__class__):
+            return (
+                self.name == other.name
+                and self.first == other.first
+                and self.length == other.length
+                and self.typeval == other.typeval
+            )
+        return False
+
+    def __repr__(self) -> str:
+        return (
+            f"Field(name={self.name}, "
+            f"typeval={repr(self.typeval)}, "
+            f"first={repr(self.first)}, "
+            f"length={repr(self.length)})"
+        )
+
+
+class Message:
+    def __init__(self, message_model: model.Message) -> None:
+        self._model = message_model
+        initial = Field(model.INITIAL.name)
+        initial.first = Number(0)
+        initial.length = Number(0)
+        self._fields: List[Field] = [initial]
+
+    def __repr__(self) -> str:
+        args = ", ".join([f"{k}={v}" for k, v in self.__dict__.items()])
+        return f"{self.__class__.__name__}({args})"
+
+    def __eq__(self, other: object) -> bool:
+        return (
+            isinstance(other, self.__class__)
+            and self._fields == other._fields
+            and self._model == other._model
+        )
+
+    def __get_field(self, fld: str) -> Field:
+        for f in self._fields:
+            if f.name == fld:
+                return f
+        raise IndexError(f"field {fld} not found")
+
+    def new(self) -> "Message":
+        m = Message(self._model)
+        return m
+
+    def __field_type(self, fld: str) -> TypeValue:
+        return TypeValue.construct(self._model.types[model.Field(fld)])
+
+    def set(self, fld: str, value: Any) -> None:
+        typedvalue = self.__field_type(fld)
+        if not isinstance(value, typedvalue.accepted_type):
+            raise TypeError(
+                f"cannot assign different types: {repr(typedvalue.accepted_type)}"
+                f"!= {repr(type(value))}"
+            )
+        typedvalue.assign(value, True)
+        incoming = self._model.incoming(model.Field(fld))
+        for i in incoming:
+            if i.source.name in [f.name for f in self._fields]:
```

Comment:
This prevents setting fields in a non-sequential way.

Reply:
Fixed by https://github.com/Componolit/RecordFlux/pull/158/commits/949c3c3aeccce59600e906a5d351d16f8de3d6ea.